tests/unit_tests/dft/include/test_common.hpp

```diff
@@ -0,0 +1,197 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef ONEMKL_TEST_COMMON_HPP
+#define ONEMKL_TEST_COMMON_HPP
+
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+template <typename T>
+struct complex_info {
+    using real_type = T;
+    static const bool is_complex = false;
+};
+
+template <typename T>
+struct complex_info<std::complex<T>> {
+    using real_type = T;
+    static const bool is_complex = true;
+};
+
+template <typename T>
+constexpr bool is_complex() {
+    return complex_info<T>::is_complex;
+}
+template <typename T>
+constexpr int num_components() {
+    return is_complex<T>() ? 2 : 1;
+}
+
+template <typename fp>
+bool check_equal(fp x, fp x_ref, int error_mag) {
+    using fp_real = typename complex_info<fp>::real_type;
+    static_assert(std::is_floating_point_v<fp_real>,
+                  "Expected floating-point real or complex type.");
+
+    const fp_real bound = [error_mag]() {
+        if constexpr (sizeof(double) == sizeof(long double) && std::is_same_v<fp_real, double>) {
+            // The reference DFT uses long double to maintain accuracy
+            // when this isn't possible, lower the accuracy requirements
+            return 1e-12;
+        }
+        else {
+            return (error_mag * num_components<fp>() * std::numeric_limits<fp_real>::epsilon());
+        }
+    }();
+
+    const auto aerr = std::abs(x - x_ref);
+    const auto rerr = aerr / std::abs(x_ref);
```

Comment:
Regarding the comments on small values of `x_ref` I think I have answered my opinion on this [above](https://github.com/oneapi-src/oneMKL/pull/261#discussion_r1092260767). I don't see the issue of having a very large `rerr`, even an infinite one, since the absolute error should be used in this case. I think having a threshold for when to use absolute error or relative error is pretty arbitrary too so this is the most elegant solution to me.

I'm not sure I follow how the relative error would be unsafe for component-wise checking. Would you suggest to remove the relative error altogether? I'm worried that large values could easily fail in that case.

Reply:
My apologies for splitting the conversation in separate comments, let's keep it going here, I'll mark the other comment as "resolved". 

Regarding the comment above on small (relative) values: for some reason I had read that the check was using a logical AND instead of a logical OR when reviewing, my bad... Thank you for the comment and for pointing out the logic in such cases.

In light of that, I believe that the current check is alright _in the current usage and for the sizes considered herein_, however I think it could become problematic if testing non-unit scaling factor and/or much larger sizes. I am certainly **not** suggesting to get rid of the relative error (please keep it!) but I believe the absolute error probably needs to be refined and/or made more robust with respect to the reference solution. I believe it currently assumes that some norm of the reference solution is order 1 but we probably want the bound on `aerr` to be proportional to that norm, in general. 

For instance, the current check on absolute error would be way too forgiving when testing a DFT with a very small scaling factor and/or with an input signal of small magnitude, instead. If we consider input data as a constant vector of values `alpha/N` where `N` is the length of the transform, then the reference output is an impulse signal of amplitude `alpha`. Now, take `alpha` much less than the current `bound`, then any backend FFT tool producing an output signal roughly bounded by `alpha` will be approved based on the check on absolute error. That includes **incorrect** solutions like an output of constant values `alpha`, an output signal as a perfect cosine wave of amplitude `alpha`, ... 
The problem in this example is that the incorrect entries of the output would be approved based on the absolute error check although the relative error does make sense in such cases and **should** be taken into account, in fact: while the output entries are indeed "small compared to 1", they are not "small compared to some norm of the reference solution" (floating point arithmetic errors do not explain the differences). Conversely, the current absolute error check becomes rather meaningless for very large `alpha` and this _may_ report false negatives in such cases, as is.

I believe this verification check would be more robust by making the bound on `aerr` proportional to "some norm" of the reference solution (my personal preference would be the max norm). Currently, the input data is random between -1 and +1 unless I'm mistaken, so `alpha` ~ `N` which could become a problem for very large `N`.

I leave it up to you whether you want to scale in the input data by `1/N` (to make the reference output roughly of order 1) or if we want to make the bound on `aerr` proportional to some norm of the reference solution, instead. My preference would be the latter.

Reply:
I have introduced a new absolute error margin to follow your suggestion in https://github.com/oneapi-src/oneMKL/pull/261/commits/925efb7168ff10def007bc378f8f8f9a7102af2e. I assumed it should still depend on the problem size as well?
Here is a log running with the MKLGPU backend to show some of the values, error margins and bounds:
[verbose_log.txt](https://github.com/oneapi-src/oneMKL/files/10560099/verbose_log.txt)

Reply:
Sorry for fragmenting the error bound discussion so much. The error bound taken from MKL examples that's a multiple of log2(N) is an absolute error bound in the context of input in [-1/N, 1/N]. I have no objections to Raphael's suggestion of absolute error bound here, but I think the relative error bound should be some small fraction right?

Reply:
The error margins are multiplied by a small epsilon [here](https://github.com/oneapi-src/oneMKL/pull/261/files#diff-a4abd74cbd5788a38db5c0928f1a46d3f2bdf5adf026b26cc6b5a1a36dba53dcR69) so the 2 bounds ends up in the range 1e-6 to 1e-15 in the current tests. Are you suggesting we should tighten more the relative error margin? There seems to be some room for a smaller margin by a factor of 10 maximum currently.

Reply:
Thank you for the changes. Can you confirm that
1. the log you attached is only for illustration purposes, i.e., temporarily disabling the `if (!ok)` here below, right? (but no actual false negative caught and reported otherwise)
2. the 1e-6 and 1e-15 range you're referring to in your last comment are the for the bounds in relative errors for single and double precision cases, respectively?
If so, I am satisfied with the suggested changes, personally.

Reply:
Yes and yes, exactly.